File: nodejs/tests/esc_schema_schema_wire_bool.test.js
// Copyright 2026, Pulumi Corporation.  All rights reserved.

// Guards EscSchemaSchema's legacy wire shorthand: JSON Schema's bare `true`/`false`
// instead of an object, meaning {always: true} / {never: true} (see
// cmd/pulumi-codegen/cmd/typescript.go emitEscSchemaSchemaFixupWire /
// emitEscSchemaSchemaWireCoercion). Mirrors
// cmd/console2/src/proxy/esc-schema-schema-wire-bool-fixup.spec.ts.

const { test } = require("node:test");
const assert = require("node:assert/strict");

const { Models } = require("../dist/index.js");

test("fixupWire coerces true into {always: true}", () => {
    const schema = Models.EscSchemaSchema.fixupWire(true);

    assert.ok(schema instanceof Models.EscSchemaSchema);
    assert.equal(schema.always, true);
    assert.ok(!schema.never);
});

test("fixupWire coerces false into {never: true}", () => {
    const schema = Models.EscSchemaSchema.fixupWire(false);

    assert.ok(schema instanceof Models.EscSchemaSchema);
    assert.equal(schema.never, true);
    assert.ok(!schema.always);
});

test("fixupWire leaves an already-object schema untouched", () => {
    const schema = Models.EscSchemaSchema.newInstance({ type: "string" });

    assert.equal(Models.EscSchemaSchema.fixupWire(schema), schema);
});

test("coerces a nested boolean schema field (ContextSchema.schema)", () => {
    const ctx = Models.ContextSchema.newInstance({ schema: true });

    assert.ok(ctx.schema instanceof Models.EscSchemaSchema);
    assert.equal(ctx.schema.always, true);
});

test("coerces boolean shorthand array elements, including the falsy `false`", () => {
    const schema = Models.EscSchemaSchema.newInstance({
        anyOf: [true, false, { type: "string" }],
    });

    assert.ok(schema.anyOf[0] instanceof Models.EscSchemaSchema);
    assert.equal(schema.anyOf[0].always, true);
    assert.ok(schema.anyOf[1] instanceof Models.EscSchemaSchema);
    assert.equal(schema.anyOf[1].never, true);
    assert.equal(schema.anyOf[2].type, "string");
});

test("coerces boolean shorthand at a map position (properties)", () => {
    const schema = Models.EscSchemaSchema.newInstance({
        properties: { foo: true, bar: false },
    });

    assert.ok(schema.properties.foo instanceof Models.EscSchemaSchema);
    assert.equal(schema.properties.foo.always, true);
    assert.ok(schema.properties.bar instanceof Models.EscSchemaSchema);
    assert.equal(schema.properties.bar.never, true);
});

test("coerces a boolean shorthand nested two levels deep", () => {
    const schema = Models.EscSchemaSchema.newInstance({
        properties: { foo: { anyOf: [true, { type: "string" }] } },
    });

    const foo = schema.properties.foo;
    assert.ok(foo instanceof Models.EscSchemaSchema);
    assert.ok(foo.anyOf[0] instanceof Models.EscSchemaSchema);
    assert.equal(foo.anyOf[0].always, true);
    assert.equal(foo.anyOf[1].type, "string");
});

File: java/src/test/java/com/pulumi/cloud/sdk/EscSchemaSchemaWireBoolTest.java
// Copyright 2026, Pulumi Corporation.  All rights reserved.

package com.pulumi.cloud.sdk;

// Hand-rolled Jackson-only test for EscSchemaSchema's legacy bare-boolean wire shorthand
// (JSON Schema's `true`/`false` for {always: true} / {never: true} instead of an object).
// See cmd/pulumi-codegen/cmd/java.go's emitEscSchemaSchemaDeserializer and Json.java's
// BeanDeserializerModifier wiring. Unlike WildcardSubtypeTest's feature, this one isn't
// Jackson-native, so this proves the generated Deserializer + Json.java plumbing actually
// round-trips through the real Json.MAPPER, both directly and nested inside another model.

import com.pulumi.cloud.sdk.model.ContextSchema;
import com.pulumi.cloud.sdk.model.EscSchemaSchema;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

class EscSchemaSchemaWireBoolTest {

    @Test
    void trueShorthandProducesAlways() throws Exception {
        EscSchemaSchema parsed = Json.MAPPER.readValue("true", EscSchemaSchema.class);

        assertTrue(parsed.always);
        assertFalse(parsed.never);
    }

    @Test
    void falseShorthandProducesNever() throws Exception {
        EscSchemaSchema parsed = Json.MAPPER.readValue("false", EscSchemaSchema.class);

        assertFalse(parsed.always);
        assertTrue(parsed.never);
    }

    @Test
    void ordinaryObjectStillDeserializesNormally() throws Exception {
        EscSchemaSchema parsed = Json.MAPPER.readValue("{\"type\":\"string\"}", EscSchemaSchema.class);

        assertEquals("string", parsed.type_);
        assertFalse(parsed.always);
        assertFalse(parsed.never);
    }

    @Test
    void shorthandResolvesWhenNestedInsideAnotherModel() throws Exception {
        ContextSchema parsed = Json.MAPPER.readValue("{\"schema\":true}", ContextSchema.class);

        assertTrue(parsed.schema.always);
    }

    @Test
    void shorthandResolvesInsideAListElement() throws Exception {
        String json = "{\"anyOf\":[true,false,{\"type\":\"string\"}]}";

        EscSchemaSchema parsed = Json.MAPPER.readValue(json, EscSchemaSchema.class);

        assertEquals(3, parsed.anyOf.size());
        assertTrue(parsed.anyOf.get(0).always);
        assertTrue(parsed.anyOf.get(1).never);
        assertEquals("string", parsed.anyOf.get(2).type_);
    }

    @Test
    void shorthandResolvesAtAMapPosition() throws Exception {
        String json = "{\"properties\":{\"foo\":true,\"bar\":false}}";

        EscSchemaSchema parsed = Json.MAPPER.readValue(json, EscSchemaSchema.class);

        assertTrue(parsed.properties.get("foo").always);
        assertTrue(parsed.properties.get("bar").never);
    }

    @Test
    void shorthandResolvesTwoLevelsDeep() throws Exception {
        String json = "{\"properties\":{\"foo\":{\"anyOf\":[true,{\"type\":\"string\"}]}}}";

        EscSchemaSchema parsed = Json.MAPPER.readValue(json, EscSchemaSchema.class);

        EscSchemaSchema foo = parsed.properties.get("foo");
        assertTrue(foo.anyOf.get(0).always);
        assertEquals("string", foo.anyOf.get(1).type_);
    }

    @Test
    void ordinaryObjectStillSerializesAsAnObject() throws Exception {
        EscSchemaSchema value = new EscSchemaSchema();
        value.type_ = "string";

        String json = Json.MAPPER.writeValueAsString(value);

        assertTrue(json.contains("\"type\":\"string\""), json);
    }
}
